Implemented postprocessor which runs after all actions on tokens finished

aleksei-berezkin opened the pull request:
The main purpose of this postprocessor is allowing doing anything with the model (or cells, or mappers) after the Hybrid Synchronizer has finished all things with tokens (complete, transform, process comments, set focus). This allows even destroying a part of a model which initiated a post-processing (see tests). Any other approach (attach model listeners, design a special completion) will inevitably cause NPE or illegal state because of changing a model while tokens and cells operations are still running.

timzam commented on hybrid/src/main/java/jetbrains/jetpad/hybrid/TokensEditPostProcessorTrait.java:
```diff
@@ -0,0 +1,47 @@
+
+class TokensEditPostProcessorTrait<SourceT> extends CellTrait {
+  private BaseHybridSynchronizer<SourceT, ?> mySync;
+  private TokensEditPostProcessor<SourceT> myPostProcessor;
```
Both fields could be final.

aleksei-berezkin replied:
Agree.

timzam commented on hybrid/src/main/java/jetbrains/jetpad/hybrid/TokensEditPostProcessorTrait.java:
```diff
@@ -0,0 +1,47 @@
+
+  @Override
+  public void onKeyPressed(Cell cell, KeyEvent event) {
+    if (isRemove(event)) {
```
Why do we have the check for isRemove() here? If you have a use case for that (when isRemove() == false) it's not covered by the tests. I removed the check and all tests passed.

aleksei-berezkin replied:
That's to exclude cases when tokens are not going to change, for example, when you hit Shift and alike, or when you type (changes will be visible in `afterTokensEdit`). I can add test which asserts tokens do change on every call.

timzam commented:
Looks good to me.

aleksei-berezkin commented:
While improving code, found several defects. Changes in last commit:
* Added postprocessing after text and tokens cut and paste.
* Added postprocessing to a placeholder token.
* Added tests to all tokens edit cases.
* `TokensEditPostProcessorTrait` renaimed to `TokenTextEditPostProcessor` because actually it's installed to `TextTokenCell` thus reacts on edits in a token. Edits in the token list are tracked in tokens container cell (one which handles copypaste).

aleksei-berezkin commented:
Discussed with Kate:
1) Unclear comment in `BaseHybridSynchronizer`
2) Use empty implementations instead of nulls

aleksei-berezkin commented:
Changes in the last commit:
1. Improved comment
1. Changed to empty implementations instead of nulls
1. Found a bug in `CompositeCellTrait`. Fix allowed tracking `ValueTokens` additions and deletions; contents editing is still not tracked (this is not needed) so comment (1) stays. New tests cover the fix.

katepol commented:
1. Now setTokensEditPostProcessor() should transform null value into empty impl
2. When we are finished, I'd like 2 commits: your stuff and bug fix at CompositeCellTrait.
Otherwise, it's OK.

aleksei-berezkin commented:
(1) done, (2) will be fixed in 2nd version with squashed commit

katepol commented:
Ok to me. @timzam ?

timzam commented:
Let's have a look at the squashed commit then.

aleksei-berezkin commented:
Squashed in the new request.
